README.md
RSpec Queue
=========================================================================
![Build status](https://github.com/skroutz/rspecq/actions/workflows/build.yml/badge.svg)
[![Gem Version](https://badge.fury.io/rb/rspecq.svg)](https://badge.fury.io/rb/rspecq)

RSpec Queue (RSpecQ) distributes and executes RSpec suites among parallel
workers. It uses a centralized queue that workers connect to and pop off
tests from. It ensures optimal scheduling of tests based on their run time,
facilitating faster CI builds.

RSpecQ is inspired by [test-queue](https://github.com/tmm1/test-queue)
and [ci-queue](https://github.com/Shopify/ci-queue).

## Features

- Run an RSpec suite among many workers
  (potentially located in different hosts) in a distributed fashion,
  facilitating faster CI builds.
- Consolidated, real-time reporting of a build's progress.
- Optimal scheduling of test execution by using timings statistics from previous runs and
  automatically scheduling slow spec files as individual examples. See
  [*Spec file splitting*](#spec-file-splitting).
- Automatic retry of test failures before being considered legit, in order to
  rule out flakiness. Additionally, flaky tests are detected and provided to
  the user. See [*Requeues*](#requeues).
- Handles intermittent worker failures (e.g. network hiccups, faulty hardware etc.)
  by detecting non-responsive workers and requeing their jobs. See [*Worker failures*](#worker-failures)
- Sentry integration for monitoring build-level events. See [*Sentry integration*](#sentry-integration).
  See [#2](https://github.com/skroutz/rspecq/issues/2).
- Automatic termination of builds after a certain amount of failures. See [*Fail-fast*](#fail-fast).

## Usage

A worker needs to be given a name and the build it will participate in.
Assuming there's a Redis instance listening at `localhost`, starting a worker
is as simple as:

```shell
$ rspecq --build=123 --worker=foo1 spec/
```

To start more workers for the same build, use distinct worker IDs but the same
build ID:

```shell
$ rspecq --build=123 --worker=foo2
```

To view the progress of the build use `--report`:

```shell
$ rspecq --build=123 --report
```

For detailed info use `--help`:

```
NAME:
    rspecq - Optimally distribute and run RSpec suites among parallel workers

USAGE:
    rspecq [<options>] [spec files or directories]

OPTIONS:
    -b, --build ID                   A unique identifier for the build. Should be common among workers participating in the same build.
    -w, --worker ID                  An identifier for the worker. Workers participating in the same build should have distinct IDs.
        --seed SEED                  The RSpec seed. Passing the seed can be helpful in many ways i.e reproduction and testing.
    -r, --redis HOST                 --redis is deprecated. Use --redis-host or --redis-url instead. Redis host to connect to (default: 127.0.0.1).
        --redis-host HOST            Redis host to connect to (default: 127.0.0.1).
        --redis-url URL              Redis URL to connect to (e.g.: redis://127.0.0.1:6379/0).
        --update-timings             Update the global job timings key with the timings of this build. Note: This key is used as the basis for job scheduling.
        --file-split-threshold N     Split spec files slower than N seconds and schedule them as individual examples.
        --report                     Enable reporter mode: do not pull tests off the queue; instead print build progress and exit when it's finished.
                                     Exits with a non-zero status code if there were any failures.
        --report-timeout N           Fail if build is not finished after N seconds. Only applicable if --report is enabled (default: 3600).
        --max-requeues N             Retry failed examples up to N times before considering them legit failures (default: 3).
        --queue-wait-timeout N       Time to wait for a queue to be ready before considering it failed (default: 30).
        --fail-fast N                Abort build with a non-zero status code after N failed examples.
        --reproduction               Enable reproduction mode: Publish files and examples in the exact order given in the command. Incompatible with --timings.
        --tag TAG                    Run examples with the specified tag, or exclude examples by adding ~ before the tag.  - e.g. ~slow  - TAG is always converted to a symbol.
    -h, --help                       Show this message.
    -v, --version                    Print the version and exit.
```

You can set most options using ENV variables:

```shell
$ RSPECQ_BUILD=123 RSPECQ_WORKER=foo1 rspecq spec/
```

### Supported ENV variables

| Name | Desc |
| --- | --- |
| `RSPECQ_BUILD` | Build ID |
| `RSPECQ_WORKER` | Worker ID |
| `RSPECQ_SEED` | RSpec seed |
| `RSPECQ_REDIS` | Redis HOST |
| `RSPECQ_UPDATE_TIMINGS` | Timings |
| `RSPECQ_FILE_SPLIT_THRESHOLD` | File split threshold |
| `RSPECQ_REPORT` | Report |
| `RSPECQ_REPORT_TIMEOUT` | Report Timeout |
| `RSPECQ_MAX_REQUEUES` | Max requests |
| `RSPECQ_QUEUE_WAIT_TIMEOUT` | Queue wait timeout |
| `RSPECQ_REDIS_URL` | Redis URL |
| `RSPECQ_FAIL_FAST` | Fail fast |
| `RSPECQ_REPORTER_RERUN_COMMAND_SKIP` | Do not report flaky test's rerun command |

### Sentry integration

RSpecQ can optionally emit build events to a
[Sentry](https://sentry.io) project by setting the
`SENTRY_DSN` environment variable.

This is convenient for monitoring important warnings/errors that may impact
build times, such as the fact that no previous timings were found and
therefore job scheduling was effectively random for a particular build.

## How it works

The core design is almost identical to ci-queue so please refer to its
[README](https://github.com/Shopify/ci-queue/blob/master/README.md) instead.

### Terminology

- **Job**: the smallest unit of work, which is usually a spec file
  (e.g. `./spec/models/foo_spec.rb`) but can also be an individual example
  (e.g. `./spec/models/foo_spec.rb[1:2:1]`) if the file is too slow.
- **Queue**: a collection of Redis-backed structures that hold all the necessary
  information for an RSpecQ build to run. This includes timing statistics,
  jobs to be executed, the failure reports and more.
- **Build**: a particular test suite run. Each build has its own **Queue**.
- **Worker**: an `rspecq` process that, given a build id, consumes jobs off the
  build's queue and executes them using RSpec
- **Reporter**: an `rspecq` process that, given a build id, waits for the build's
  queue to be drained and prints the build summary report

### Spec file splitting

Particularly slow spec files may set a limit to how fast a build can be.
For example, a single file may need 10 minutes to run while all other
files finish after 8 minutes. This would cause all but one workers to be
sitting idle for 2 minutes.

To overcome this issue, RSpecQ can split files which their execution time is
above a certain threshold (set with the `--file-split-threshold` option)
and instead schedule them as individual examples.

Note: In the future, we'd like for the slow threshold to be calculated and set
dynamically (see #3).

### Requeues

As a mitigation technique against flaky tests, if an example fails it will be
put back to the queue to be picked up by another worker. This will be repeated
up to a certain number of times (set with the `--max-requeues` option), after
which the example will be considered a legit failure and printed as such in the
final report.

Flaky tests are also detected and printed as such in the final report. They are
also emitted to Sentry (see [Sentry integration](#sentry-integration)).

### Fail-fast

In order to prevent large suites running for a long time with a lot of
failures, a threshold can be set to control the number of failed examples that
will render the build unsuccessful. This is in par with RSpec's
[--fail-fast](https://relishapp.com/rspec/rspec-core/docs/command-line/fail-fast-option).

This feature is disabled by default, and can be controlled via the
`--fail-fast` command line option.

### Worker failures

It's not uncommon for CI processes to encounter unrecoverable failures for
various reasons: faulty hardware, network hiccups, segmentation faults in
MRI etc.

For resiliency against such issues, workers emit a heartbeat after each
example they execute, to signal
that they're healthy and performing jobs as expected. If a worker hasn't
emitted a heartbeat for a given amount of time (set by `WORKER_LIVENESS_SEC`)
it is considered dead and its reserved job will be put back to the queue, to
be picked up by another healthy worker.


## Rationale

### Why didn't you use ci-queue?

**Update**: ci-queue [deprecated support for RSpec](https://github.com/Shopify/ci-queue/pull/149).

While evaluating ci-queue we experienced slow worker boot
times (up to 3 minutes in some cases) combined with disk IO saturation and
increased memory consumption. This is due to the fact that a worker in
ci-queue has to load every spec file on boot. In applications with a large
number of spec files this may result in a significant performance hit and
in case of cloud environments, increased costs.

We also observed slower build times compared to our previous solution which
scheduled whole spec files (as opposed to individual examples), due to
big differences in runtimes of individual examples, something common in big
RSpec suites.

We decided for RSpecQ to use whole spec files as its main unit of work (as
opposed to ci-queue which uses individual examples). This means that an RSpecQ
worker only loads the files needed and ends up with a subset of all the suite's
files.  (Note: RSpecQ also schedules individual examples, but only when this is
deemed necessary, see [Spec file splitting](#spec-file-splitting)).

This kept boot and test run times considerably fast. As a side benefit, this
allows suites to keep using `before(:all)` hooks (which ci-queue explicitly
rejects).

The downside of this design is that it's more complicated, since the scheduling
of spec files happens based on timings calculated from previous runs. This
means that RSpecQ maintains a key with the timing of each job and updates it
on every run (if the `--update-timings` option was used). Also, RSpecQ has a
"slow file threshold" which, currently has to be set manually (but this can be
improved in the future).


## Development

Install the required dependencies:

```
$ bundle install
```

Then you can execute the tests after spinning up a Redis instance at
`127.0.0.1:6379`:

```
$ bundle exec rake
```

To enable verbose output in the tests:

```
$ RSPECQ_DEBUG=1 bundle exec rake
```

## Redis

RSpecQ by design doesn't expire its keys from Redis. It is left to the user
to configure the Redis server to do so; see
[Using Redis as an LRU cache](https://redis.io/topics/lru-cache) for more info.

You can do this from a configuration file or with `redis-cli`.

## License

RSpecQ is licensed under MIT. See [LICENSE](LICENSE).


CHANGELOG.md
# Changelog

Breaking changes are prefixed with a "[BREAKING]" label.

## master (unreleased)

## 0.7.2 (2021-11-15)

- Add tag option to filter specs (@jorge-wonolo)

## 0.7.1 (2021-04-08)

- New env variable RSPECQ_REPORTER_RERUN_COMMAND_SKIP. When set, the reporter
  does not include the flaky test's rerun command.

## 0.7.0 (2021-04-01)

- New cli parameter `reproduction`.
  When passed, primary worker publishes the queue in the same order as passed
  in the command.
- Reporter now includes a reproduction command for flaky tests.

## 0.6.0 (2021-03-23)

- New cli parameter `seed`.
  The seed is passed to the RSpec command.

## 0.5.0 (2021-02-05)

### Added

- New cli parameter `queue_wait_timeout`.
  It configured the time a queue can wait to be ready. The env equivalent
  is `RSPECQ_QUEUE_WAIT_TIMEOUT`. [#51](https://github.com/skroutz/rspecq/pull/51)

## 0.4.0 (2020-10-07)

### Added

- Builds can be configured to terminate after a specified number of failures,
  using the `--fail-fast` option.


## 0.3.0 (2020-10-05)

### Added

- Providing a Redis URL is now possible using the `--redis-url` option
  [[#40](https://github.com/skroutz/rspecq/pull/40)]

### Changed

- [DEPRECATION] The `--redis` option is now deprecated. Use `--redis-host`
  instead [[#40](https://github.com/skroutz/rspecq/pull/40)]

## 0.2.2 (2020-09-10)

### Fixed
- Worker would fail if application code was writing to stderr
 [[#35](https://github.com/skroutz/rspecq/pull/35)]

## 0.2.1 (2020-09-09)

### Changed

- Sentry Integration: Changed the way events for flaky jobs are emitted to a
  per-flaky-job fashion. This ultimately improves grouping and filtering of the
  flaky events in Sentry [[#33](https://github.com/skroutz/rspecq/pull/33)]


## 0.2.0 (2020-08-31)

This is a feature release with no breaking changes.

### Added

- Flaky jobs are now printed by the reporter in the final build output and also
  emitted to Sentry (if the integration is enabled) [[#26](https://github.com/skroutz/rspecq/pull/26)]

## 0.1.0 (2020-08-27)

### Added

- Sentry integration for various RSpecQ-level events [[#16](https://github.com/skroutz/rspecq/pull/16)]
- CLI: Flags can now be also set environment variables [[c519230](https://github.com/skroutz/rspecq/commit/c5192303e229f361e8ac86ae449b4ea84d42e022)]
- CLI: Added shorthand specifiers versions for some flags [[df9faa8](https://github.com/skroutz/rspecq/commit/df9faa8ec6721af8357cfee4de6a2fe7b32070fc)]
- CLI: Added `--help` and `--version` flags [[df9faa8](https://github.com/skroutz/rspecq/commit/df9faa8ec6721af8357cfee4de6a2fe7b32070fc)]
- CLI: Max number of retries for failed examples is now configurable via the `--max-requeues` option [[#14](https://github.com/skroutz/rspecq/pull/14)]

### Changed

- [BREAKING] CLI: Renamed `--timings` to `--update-timings` [[c519230](https://github.com/skroutz/rspecq/commit/c5192303e229f361e8ac86ae449b4ea84d42e022)]
- [BREAKING] CLI: Renamed `--build-id` to `--build` and `--worker-id` to `--worker` [[df9faa8](https://github.com/skroutz/rspecq/commit/df9faa8ec6721af8357cfee4de6a2fe7b32070fc)]
- CLI: `--worker` is not required when `--reporter` is used [[4323a75](https://github.com/skroutz/rspecq/commit/4323a75ca357274069d02ba9fb51cdebb04e0be4)]
- CLI: Improved help output [[df9faa8](https://github.com/skroutz/rspecq/commit/df9faa8ec6721af8357cfee4de6a2fe7b32070fc)]


lib/rspecq/parser.rb
require "optparse"

module RSpecQ
  class Parser
    DEFAULT_REDIS_HOST = "127.0.0.1".freeze
    DEFAULT_REPORT_TIMEOUT = 3600 # 1 hour
    DEFAULT_MAX_REQUEUES = 3
    DEFAULT_QUEUE_WAIT_TIMEOUT = 30
    DEFAULT_FAIL_FAST = 0

    def self.parse!(args)
      new(args).parse!
    end

    attr_reader :args, :opts

    def initialize(args)
      @args = args
      @opts = {}
    end

    # This method mutates `args` in order to allow both rspecq and
    # rspec options to be passed to rspecq. ["--build", "foo",
    # "--", "--pattern", "bar"] will set `build: "foo"` for rspecq
    # options and leave ["--pattern", "bar"] to be passed to rspec
    def parse!
      parse_args!
      parse_env

      # rubocop:disable Style/RaiseArgs, Layout/EmptyLineAfterGuardClause
      raise OptionParser::MissingArgument.new(:build) if opts[:build].nil?
      raise OptionParser::MissingArgument.new(:worker) if !opts[:report] && opts[:worker].nil?
      # rubocop:enable Style/RaiseArgs, Layout/EmptyLineAfterGuardClause

      opts
    end

    private

    def parse_args!
      OptionParser.new do |o|
        name = File.basename($PROGRAM_NAME)

        o.banner = <<~BANNER
          NAME:
              #{name} - Optimally distribute and run RSpec suites among parallel workers

          USAGE:
              #{name} [<options>] [spec files or directories]
        BANNER

        o.separator ""
        o.separator "OPTIONS:"

        o.on("-b", "--build ID", "A unique identifier for the build. Should be " \
            "common among workers participating in the same build.") do |v|
          opts[:build] = v
        end

        o.on("-w", "--worker ID", "An identifier for the worker. Workers " \
            "participating in the same build should have distinct IDs.") do |v|
          opts[:worker] = v
        end

        o.on("--seed SEED", "The RSpec seed. Passing the seed can be helpful in " \
          "many ways i.e reproduction and testing.") do |v|
          opts[:seed] = v
        end

        o.on("-r", "--redis HOST", "Redis host to connect to " \
            "(default: #{DEFAULT_REDIS_HOST}).") do |v|
          puts "--redis is deprecated. Use --redis-host or --redis-url instead"
          opts[:redis_host] = v
        end

        o.on("--redis-host HOST", "Redis host to connect to " \
            "(default: #{DEFAULT_REDIS_HOST}).") do |v|
          opts[:redis_host] = v
        end

        o.on("--redis-url URL", "The URL of the Redis host to connect to " \
            "(e.g.: redis://127.0.0.1:6379/0).") do |v|
          opts[:redis_url] = v
        end

        o.on("--update-timings", "Update the global job timings key with the "     \
            "timings of this build. Note: This key is used as the basis for job " \
            "scheduling.") do |v|
          opts[:timings] = v
        end

        o.on("--file-split-threshold N", Integer, "Split spec files slower than N " \
            "seconds and schedule them as individual examples.") do |v|
          opts[:file_split_threshold] = v
        end

        o.on("--report", "Enable reporter mode: do not pull tests off the queue; " \
                        "instead print build progress and exit when it's "        \
                        "finished.\n#{o.summary_indent * 9} "                       \
                        "Exits with a non-zero status code if there were any "    \
                        "failures.") do |v|
          opts[:report] = v
        end

        o.on("--report-timeout N", Integer, "Fail if build is not finished after " \
            "N seconds. Only applicable if --report is enabled "                  \
            "(default: #{DEFAULT_REPORT_TIMEOUT}).") do |v|
          opts[:report_timeout] = v
        end

        o.on("--max-requeues N", Integer, "Retry failed examples up to N times "   \
            "before considering them legit failures "                             \
            "(default: #{DEFAULT_MAX_REQUEUES}).") do |v|
          opts[:max_requeues] = v
        end

        o.on("--queue-wait-timeout N", Integer, "Time to wait for a queue to be "   \
            "ready before considering it failed "                                  \
            "(default: #{DEFAULT_QUEUE_WAIT_TIMEOUT}).") do |v|
          opts[:queue_wait_timeout] = v
        end

        o.on("--fail-fast N", Integer, "Abort build with a non-zero status code " \
            "after N failed examples.") do |v|
          opts[:fail_fast] = v
        end

        o.on("--reproduction", "Enable reproduction mode: run rspec on the given files " \
            "and examples in the exact order they are given. Incompatible with " \
            "--timings.") do |v|
          opts[:reproduction] = v
        end

        o.on("--junit-output filepath", String, "Output junit formatted xml "    \
            "for CI suites to the defined file path. Substitution parameters "  \
            "{{TEST_ENV_NUMBER}} - parallel gem proc number. "                  \
            "{{JOB_INDEX}} - increments with each suite that is run.") do |v|
          opts[:junit_output] = v
        end

        o.on("--include-pattern PATTERN", "Run tests matching the regex") do |v|
          opts[:include_pattern] = /#{v}/
        end

        o.on("--exclude-pattern PATTERN", "Regex of tests to exclude from run.") do |v|
          opts[:exclude_pattern] = /#{v}/
        end

        o.on_tail("-h", "--help", "Show this message.") do
          puts o
          exit
        end

        o.on_tail("-v", "--version", "Print the version and exit.") do
          puts "#{name} #{RSpecQ::VERSION}"
          exit
        end
      end.parse!(args)
    end

    def parse_env
      opts[:build] ||= ENV["RSPECQ_BUILD"]
      opts[:worker] ||= ENV["RSPECQ_WORKER"]
      opts[:seed] ||= ENV["RSPECQ_SEED"]
      opts[:redis_host] ||= ENV["RSPECQ_REDIS"] || DEFAULT_REDIS_HOST
      opts[:timings] = opts.fetch(:timings, env_set?("RSPECQ_UPDATE_TIMINGS"))
      opts[:file_split_threshold] ||= Integer(ENV["RSPECQ_FILE_SPLIT_THRESHOLD"] || 9_999_999)
      opts[:report] = opts.fetch(:report, env_set?("RSPECQ_REPORT"))
      opts[:report_timeout] ||= Integer(ENV["RSPECQ_REPORT_TIMEOUT"] || DEFAULT_REPORT_TIMEOUT)
      opts[:max_requeues] ||= Integer(ENV["RSPECQ_MAX_REQUEUES"] || DEFAULT_MAX_REQUEUES)
      opts[:queue_wait_timeout] ||= Integer(ENV["RSPECQ_QUEUE_WAIT_TIMEOUT"] || DEFAULT_QUEUE_WAIT_TIMEOUT)
      opts[:redis_url] ||= ENV["RSPECQ_REDIS_URL"]
      opts[:fail_fast] ||= Integer(ENV["RSPECQ_FAIL_FAST"] || DEFAULT_FAIL_FAST)
      opts[:reproduction] ||= env_set?("RSPECQ_REPRODUCTION")
      opts[:junit_output] ||= ENV["RSPECQ_JUNIT_OUTPUT"]
      opts[:include_pattern] ||= ENV["INCLUDE_PATTERN"]
      opts[:exclude_pattern] ||= ENV["EXCLUDE_PATTERN"]
    end

    def env_set?(var)
      ["1", "true"].include?(ENV[var])
    end
  end
end
